### DISCUSSION TITLE
Utilizing Unified Memory for Oceananigans simulations

### DISCUSSION THREAD

**Discussion by BrodiePearson** [CliMA/Oceananigans.jl#3592] [Experimental features]
**Would it be helpful/appropriate to add some information about _Unified Memory_ utilization in the Simulation Tips/GPU section?** If so, I'd be happy to draft a short new bullet point for the "decrease memory use of your runs" section of the documentation.

I've recently run Oceananigans non-hydrostatic (3D isotropic) simulations utilizing _Unified Memory_ on new NVIDIA hardware (GH200; Grace Hopper Superchips). _Unified Memory_ allows bigger simulations on a single GPU (>1 billion grid points in this case), compared to using the GPUs _Device Memory_ (which is the on-GPU RAM we typically use), but it comes at the cost of a slower simulation due to interconnect bandwidth. Some preliminary simulations on the GH200 (see below) are 2-3x larger but are up to 3x slower than expected by extrapolating _Device Memory_ scalings.

As background:

- The GH200 architecture has GPUs and CPUs integrated on a single chip, with 96GB of GPU _Device Memory_, and 480GB of _Unified Memory_  which can be quickly accessed by both CPUs and the GPU (albeit not as quick-access as the GPU's own memory pool). 
- Having both unified and device memory seems to be the direction NVIDIA is moving (Grace Blackwell GB200, etc.), and is also the structure used by Apple Silicon (although as far as I understand it, CUDA does not work on Apple Silicon), so this functionality may become more generally useful soon.
- Although I expect GPU-parallelization is the route to the most complex simulations, _Unified Memory_ utilization could help users who only have access to single-GPUs or GPU arrays with low-bandwidth interconnects. 
- To utilize _Unified Memory_, CUDA.jl has recently-added functionality to adjust the default memory utilized for CUDA Arrays etc. After lots of trial and error, I found that I just had to find my Julia environment's _LocalPreferences.toml_ file and add the following lines to it:

```
[CUDA]
default_memory = "unified"
```

![3D_Isotropic_Device_vs_Unified_Memory](https://github.com/CliMA/Oceananigans.jl/assets/43418592/f03a17db-1e8b-488d-afc0-897b9b30639a){width=50}

**Comment by glwagner** [comment]
That is really interesting!

> I found that I just had to find my Julia environment's LocalPreferences.toml file and add the following lines to it:

What specifically does that line do? Is there documentation for `CUDA.jl` that you can share? Also to clarify, if this is included, then the simulation will run more slowly in cases that don't require unified memory, is that right?

> is also the structure used by Apple Silicon (although as far as I understand it, CUDA does not work on Apple Silicon)

CUDA is specific to NVidia hardware. For Apple Silicon, we can use [`Metal.jl`](https://github.com/JuliaGPU/Metal.jl) (which is supported by KernelAbstractions. There's been a little work to get this working #3288. But it doesn't seem to be a priority for anyone --- could be an interesting side or undergraduate project, though. It's probably not too much work, but does require a little spin up to become familiar with writing Julia extensions and also figuring out the necessary translations in Oceananigans source code (which are not many).

It would be interesting to test the hydrostatic model in a configuration that doesn't have FFTs and to compare performance between unified memory and explicit parallelization. @simone-silvestri might be interested in that.

**Reply by BrodiePearson** [reply]
> > I found that I just had to find my Julia environment's LocalPreferences.toml file and add the following lines to it:

>  What specifically does that line do? Is there documentation for CUDA.jl that you can share?

That line changes the memory that CUDA allocates by default (for example when creating arrays). There is [CUDA.jl documentation here](https://cuda.juliagpu.org/stable/usage/memory/#Unified-memory) that describes how to create a single CUDA array within the Unified memory rather than Device memory. The last sentence of the linked section mentions that you can change the `default_memory` to `unified`, which is what you need to do to run software where it is unfeasible to modify every CUDA object manually. It doesn't describe how to do that, but `default_memory` is a constant, so it must be changed in advance of initailizing CUDA in your Julia instance (and I found the clue on [how to do that here](https://github.com/JuliaGPU/CUDA.jl/blob/64b1c486b1f95c80930114047d8cc232ca8bd5b4/LocalPreferences.toml#L17)).

> Also to clarify, if this is included, then the simulation will run more slowly in cases that don't require unified memory, is that right?

Correct, **you only want this default setting active when trying to run something that requires the expanded _Unified memory_ pool**. Otherwise you should have it commented out or set to `device`. Unfortunately, you cannot just redefine `default_memory` in a given Julia instance, as it is a constant (see above). Maybe there is a way to call `using CUDA` in a way that specifies `default_memory` but that is beyond my knowledge.

I assumed as much with Apple Silicon (which is not something I'm looking at, but saw an old Unified memory discussion on here that mentioned using those chips).

I'd be happy to try a hydrostatic setup. I only envisage there being a small overlap of simulations that benefit from either this or explicit parallelization (which can scale much larger), but it would be interesting to compare within that range!

**Reply by glwagner** [reply]
Hmm, well it looks like you can also manually construct `CuArray` with the unified backend. So it might be possible to design an abstraction that allows the memory type to be specified in `GPU`, eg `GPU(unified_memory=true)` or something. But this seems niche enough for now that we don't need to worry much yet.

> I only envisage there being a small overlap of simulations that benefit from either this or explicit parallelization (which can scale much larger), but it would be interesting to compare within that range!

True.

**Reply by BrodiePearson** [reply]
> Hmm, well it looks like you can also manually construct CuArray with the unified backend. So it might be possible to design an abstraction that allows the memory type to be specified in GPU, eg GPU(unified_memory=true) or something.

That would be nice to build in to the Architecture module. However, there is currently behavior that seems strange (at least to me; perhaps due to naivety) where a CuArray will revert to CUDA's default_memory (typically Device Memory) after an arithmetic operation, even if the initial CuArray and the operand were allocated to different memory (see below). I'm guessing that might cause issues...

![Unified_to_Device_Memory_Transition](https://github.com/CliMA/Oceananigans.jl/assets/43418592/a558e3b7-04d2-4e64-acad-dc4bd2513d8a)

**Reply by glwagner** [reply]
Oh I see, that's interesting. So even though `b` has the `UnifiedBuffer` backend, when CUDA allocates _new_ memory because we wrote `2 * b`, it will allocate that new memory using the default buffer rather than propagating the buffer in `b`.

That might be what the docs mean with this:

> Right now, CUDA.jl still defaults to allocating device memory, but this may change in the future. If you want to change the default behavior, you can set the default_memory preference to unified or host instead of device.

from here: https://cuda.juliagpu.org/stable/usage/memory/#Unified-memory

I'm not sure that would affect us --- I think that we allocate all of our memory intentionally / in a function, and we don't rely on implicit allocation via operations like `2 * b`. But it would be a good test that this is indeed true...

A bit more...

`default_memory` is a const:

https://github.com/JuliaGPU/CUDA.jl/blob/cdd80785b4e3cd26be15b54758d3fa9c2fb4ba01/src/array.jl#L117-L123

There may a be a way to set preferences programmatically through `Preferences.jl` `@set_preferences` macro (or `set_preferences!`): https://juliapackaging.github.io/Preferences.jl/stable/reference/#Preferences.set_preferences!

**Reply by glwagner** [reply]
I think possibly at the top of a script, before writing `using CUDA`, you could try writing something like

```julia
set_preferences!("CUDA", "default_memory" => "unified", force=true)
```

(`force=true` is needed to override a default it looks like) I believe we can also support loading a preference through an environment variable in Oceananigans (if we want) which would mean people could do something like

```
OCEANANIGANS_MEMORY=unified julia --project run_my_script.jl
```

for example

**Reply by glwagner** [reply]
the environment variable thing could also be implemented in your script I guess with something like

```julia
using Preferences

cuda_default_memory = ENV["DEFAULT_MEMORY"]
set_preferences!("CUDA", "default_memory" => cuda_default_memory)

using Oceananigans
# etc
```

**Reply by glwagner** [reply]
Hmm, the downside of this approach is that the preference will then be used for _all_ subsequent simulations.

So I think it's better to use a `LocalPreferences.toml` file.

I just want to clarify this:

> I found that I just had to find my Julia environment's LocalPreferences.toml file and add the following lines to it:

Just to clarify, if we always use a local environment with a `Project.toml` for all the scripts we run, then we don't have to "find" a `LocalPreferences.toml` file, right? What we should be doing instead is to _create_ a _new_ `LocalPreferences.toml` within our current environment (rather than using a global environment), and then run our script with `julia --project`. 

I think that usage is good because typically one should have different environments for different experiments. So an experiment that is big enough to need unified memory can have its custom environment reflect that.

**Reply by glwagner** [reply]
I put together a repo to illustrate how to do this:

https://github.com/glwagner/ReallyBigSimulations/tree/main

**Reply by BrodiePearson** [reply]
@glwagner Great, that repo is a nice concise example (although `this_is_big.jl` may be a misnomer haha).

Perhaps something to the effect of this could be added to the bullet points in [this section of the documentation](https://clima.github.io/OceananigansDocumentation/stable/simulation_tips/#Try-to-decrease-the-memory-use-of-your-runs)?

> Some hardware includes a `unified memory` pool shared between both GPU(s) and CPUs, which can be larger than the GPU `device memory` that we typically use for single-GPU computations. If `unified memory` is available, you may be able to conduct a larger  simulation than is possible with the `device memory` (for example, grids a bit larger than 1024 x 1024 x 1024 are possible on a Grace Hopper Superchip with 480GB unified memory). To access `unified memory` for an Oceananigans simulation, you can use a `LocalPreferences.toml` file as described in [this example](https://github.com/glwagner/ReallyBigSimulations/tree/main).  It is not recommended to use `unified memory` for problems that could fit on `device memory` because the GPU and `unified memory` typically have a lower-bandwidth interconnect than the GPU and `device memory`.

Also, 

> I'm not sure that would affect us --- I think that we allocate all of our memory intentionally / in a function, and we don't rely on implicit allocation via operations like 2 * b. But it would be a good test that this is indeed true...

That's good to know, I knew that each array is initially allocated to the chosen memory, but didn't realize this memory allocation is done again when arrays are manipulated (in my example, `c` was initially a Unified Memory array, but overwriting it without specifying memory changed it to `default_memory`).

**Reply by glwagner** [reply]
> (in my example, c was initially a Unified Memory array, but overwriting it without specifying memory changed it to default_memory).

Yes but let me clarify. You didn't overwrite the _array_, you just overwrote the _name_ `c`. More precisely, two things occur if you write `c = 2 * a`. First, `2 * a` has to be computed and stored somewhere. So memory is allocated, and apparently, it used the default memory. The next step, which really is a distinct step, is to make the name `c` _point_ to that newly allocated memory. You could have also written `c .= 2 .* a`. But because `c=a` before this line, this computation would actually be stored in the memory space that `a` points to --- which may not have been your intent. For example:

```julia
julia> a = rand(3, 3)
3×3 Matrix{Float64}:
 0.0646543  0.869536   0.485728
 0.929745   0.100563   0.34596
 0.720914   0.0197468  0.792766

julia> c = a
3×3 Matrix{Float64}:
 0.0646543  0.869536   0.485728
 0.929745   0.100563   0.34596
 0.720914   0.0197468  0.792766

julia> c .= 2 .* a
3×3 Matrix{Float64}:
 0.129309  1.73907    0.971457
 1.85949   0.201125   0.691919
 1.44183   0.0394937  1.58553

julia> a
3×3 Matrix{Float64}:
 0.129309  1.73907    0.971457
 1.85949   0.201125   0.691919
 1.44183   0.0394937  1.58553
```

Notice that `a` is now different. That's because `c .= 2 .* a` is exactly the same as `a .= 2 .* a` or `a .*= 2`.

To exert more control over memory allocation, you might try this pattern instead:

```julia
julia> a = rand(3, 3)
3×3 Matrix{Float64}:
 0.228287  0.263259  0.852691
 0.787519  0.697647  0.939554
 0.126515  0.684663  0.970198

julia> b = similar(a) # create a new array that is "similar" to a. deepcopy(a) probably works too
3×3 Matrix{Float64}:
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

julia> b .= 2 .* a # broadcast the operation 2 * a over all the elements of a and store in b
3×3 Matrix{Float64}:
 0.456573  0.526517  1.70538
 1.57504   1.39529   1.87911
 0.253029  1.36933   1.9404
```

Notice the `.` they are very important. And then if you try

```julia
julia> c = 2 .* a # create a new array by computing 2 * a and assign the name `c` to that array
3×3 Matrix{Float64}:
 0.456573  0.526517  1.70538
 1.57504   1.39529   1.87911
 0.253029  1.36933   1.9404

julia> b === c # b and c point to different underlying memory (even though they are numerically equal)
false
```

**Reply by BrodiePearson** [reply]
Thanks, that makes more sense now and I see what I was doing wrong - those operations, with appropriately placed dots and the similar function, maintain the unified memory location of the final array.

**Reply by glwagner** [reply]
I think all memory allocation passes through this line in fact:

https://github.com/CliMA/Oceananigans.jl/blob/fb2c670626e38e3bffe298c485f95625fb1d83be/src/Grids/zeros_and_ones.jl#L7

which has the architecture. So adding a `memory` to `GPU` (for `Unified()` or `Device()`) might really be all that's needed.
